fastembed/embedding.py

```diff
@@ -304,75 +405,48 @@ def __init__(
-            if optimized_model_path.exists():
-                optimized_model_path.rename(model_path)
-            else:
-                raise ValueError(f"Could not find model.onnx in {model_dir}")
-
-        # Hacky support for multilingual model
-        self.exclude_token_type_ids = False
-        if model_name == "intfloat/multilingual-e5-large":
-            self.exclude_token_type_ids = True
-
-        so = ort.SessionOptions()
-        so.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
-
-        self.tokenizer = self.load_tokenizer(model_dir, max_length=max_length)
-
-        self.model = ort.InferenceSession(str(model_path), providers=onnx_providers, sess_options=so)
-
-    def onnx_embed(self, documents: List[str]) -> np.ndarray:
-        encoded = self.tokenizer.encode_batch(documents)
-        input_ids = np.array([e.ids for e in encoded])
-        attention_mask = np.array([e.attention_mask for e in encoded])
-
-        onnx_input = {
-            "input_ids": np.array(input_ids, dtype=np.int64),
-            "attention_mask": np.array(attention_mask, dtype=np.int64),
-        }
-
-        if not self.exclude_token_type_ids:
-            onnx_input["token_type_ids"] = np.array(
-                [np.zeros(len(e), dtype=np.int64) for e in input_ids], dtype=np.int64
-            )
-
-        model_output = self.model.run(None, onnx_input)
-        last_hidden_state = model_output[0][:, 0]
-        embeddings = normalize(last_hidden_state).astype(np.float32)
-        return embeddings
-
-    def embed(self, documents: List[str], batch_size: int = 256) -> Iterable[np.ndarray]:
+    def embed(
+            self, documents: Union[str, Iterable[str]], batch_size: int = 256, parallel: int = 0
+    ) -> Iterable[np.ndarray]:
         """
         Encode a list of documents into list of embeddings.
         We use mean pooling with attention so that the model can handle variable-length inputs.
 
         Args:
-            documents: List of documents to embed
+            documents: Iterator of documents or single document to embed
             batch_size: Batch size for encoding -- higher values will use more memory, but be faster
+            parallel:
+                If > 1, data-parallel encoding will be used, recommended for offline encoding of large datasets.
+                If 0, will use default threading configuration.
 
         Returns:
             List of embeddings, one per document
         """
-        if type(documents) == str:
+        if isinstance(documents, str):
             documents = [documents]
-        # TODO: Replace loop with parallelized batching
-        if len(documents) >= batch_size:
-            for i in range(0, len(documents), batch_size):
-                batch = documents[i : i + batch_size]
-                yield from self.onnx_embed(batch)
+        if parallel == 0:
+            parallel = os.cpu_count()
+
+        if parallel == 0:
```

Comment:
Should we start multiple threads when the batch count is larger than some given threshold? E.g. for 1-5 batches we use a single process? Wondering if there is a trade off to be made when we've exactly one sentence, one batch to embed i.e. query time.